frankenstein/server.go

```diff
@@ -20,16 +20,24 @@ import (
 	"github.com/golang/protobuf/proto"
 	"github.com/prometheus/common/log"
 	"github.com/prometheus/common/model"
+	"golang.org/x/net/context"
 
-	"github.com/prometheus/prometheus/storage"
 	"github.com/prometheus/prometheus/storage/metric"
 	"github.com/prometheus/prometheus/storage/remote/generic"
 )
 
+// SampleAppender is the interface to append samples to both, local and remote
+// storage. All methods are goroutine-safe.
+type SampleAppender interface {
+	Append(context.Context, *model.Sample) error
+	NeedsThrottling(context.Context) bool
+}
+
 // AppenderHandler returns a http.Handler that accepts protobuf formatted
 // metrics and sends them to the supplied appender.
-func AppenderHandler(appender storage.SampleAppender) http.Handler {
+func AppenderHandler(appender SampleAppender) http.Handler {
 	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		ctx := context.WithValue(context.Background(), UserIDContextKey, r.Header.Get(UserIDContextKey))
```

Comment:
Check and return an error here if the header is not present?

Further down the stack you check whether the context has the user ID field at all, but it could still be empty (e.g. if the upstream proxy screws up and doesn't set the header) - in the unlucky case I think this could mean that all users accidentally share their metrics with each other instead of an error.